File: src/java/org/voegtle/weatherstation/server/ImageServlet.java
package org.voegtle.weatherstation.server;

import org.voegtle.weatherstation.server.image.Image;
import org.voegtle.weatherstation.server.image.ImageCache;
import org.voegtle.weatherstation.server.persistence.ImageIdentifier;
import org.voegtle.weatherstation.server.request.ImageUrlParameter;
import org.voegtle.weatherstation.server.util.StringUtil;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class ImageServlet extends AbstractServlet {
  private ImageCache imageCache;

  @Override
  public void init() throws ServletException {
    super.init();
    imageCache = new ImageCache(pm);
    imageCache.init();
  }

  @Override
  protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {

    ImageUrlParameter param = new ImageUrlParameter(request);
    if (param.isRefresh()) {
      imageCache.refresh(param.getBegin(), param.getEnd());
      response.getWriter().write("ACK " + imageCache.size());
      response.getWriter().close();
    } else if (param.isClear()) {
      imageCache.clear();
      response.getWriter().write("ACK");
      response.getWriter().close();
    } else {
      if (StringUtil.isNotEmpty(param.getOid())) {
        ImageIdentifier identifier;
        if (StringUtil.isNotEmpty(param.getZx())) {
          identifier = new ImageIdentifier(param.getOid(), param.getZx());
        } else {
          identifier = new ImageIdentifier(param.getSheet(), param.getOid(), param.getFormat());
        }
        Image image = imageCache.get(identifier);

        if (image != null) {
          response.setContentType("image/png");
          response.getOutputStream().write(image.getPngAsBytes());
          response.getOutputStream().close();
        }
      }
    }

  }


}

File: src/main/java/org/voegtle/weatherstation/server/ImageServlet.kt
package org.voegtle.weatherstation.server

import org.voegtle.weatherstation.server.image.ImageCache
import org.voegtle.weatherstation.server.persistence.entities.ImageIdentifier
import org.voegtle.weatherstation.server.request.ImageUrlParameter
import java.io.IOException
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class ImageServlet : AbstractServlet() {
  private val imageCache = ImageCache(pm)

  @Throws(ServletException::class)
  override fun init() {
    super.init()
    runInObjectifyContext {
      val imageIdentifier = ImageIdentifier(1, "92028519")
      pm.makePersistent(imageIdentifier)

      imageCache.init()
    }
  }

  @Throws(ServletException::class, IOException::class)
  override fun doGet(request: HttpServletRequest, response: HttpServletResponse) {

    val param = ImageUrlParameter(request, locationProperties!!.dateUtil)
    when {
      param.isRefresh -> {
        imageCache.refresh(param.begin, param.end)
        response.writer.write("ACK ${imageCache.size()}")
        response.writer.close()
      }
      param.isClear -> {
        imageCache.clear()
        response.writer.write("ACK")
        response.writer.close()
      }
      else -> param.oid?.let {
        val identifier = ImageIdentifier(param.sheet, it)

        val image = imageCache[identifier]

        image?.let {
          response.contentType = "image/png"
          response.outputStream.write(image.pngAsBytes)
          response.outputStream.close()
        }
      }
    }

  }


}
